readme.md
![tests](https://github.com/JGoutin/compilertools/workflows/tests/badge.svg)
[![codecov](https://codecov.io/gh/JGoutin/compilertools/branch/master/graph/badge.svg)](https://codecov.io/gh/JGoutin/compilertools)
[![PyPI](https://img.shields.io/pypi/v/compilertools.svg)](https://pypi.org/project/compilertools)

Python uses the Wheel format for simplified package distribution. However,
it does not allow distributing packages optimized for each machine but highly compatible ones.
The user must compile the package himself to take advantage of optimization like SIMD (SSE, AVX, FMA, ...).

Compilertools allows working around this problem and distributing optimized packages for several machines while keeping
the simplicity of Wheel. It works in the background and has been created with the aim of being easy to use.
Package maintainer requires only importing it at runtime and buildtime. Everything is transparent for the end user.

Its secondary objective is also to help the package maintainer to optimally compile its package with multiple compilers
by configuring options for him.

Documentation:
--------------
[**Compilertools Documentation**](https://jgoutin.github.io/compilertools/)


Features:
---------

* Compiles and distributes optimized binaries for a variety of machines in a single Wheel package.
* Helps to build optimized package from sources for current machine.
* Handles automatically compiling and linking options for a variety of compilers.
* Autodetects openMP, OpenACC or Intel Cilk Plus in source code and automatically sets related compiling and linking
  options.
* Support extra compiling options like fast math.
* Provides build time settings for package maintainer to tweak compilation.
* Provides API for getting information on current machine CPU.
* Lightweight pure Python module with no dependency that uses lazy import and evaluation as possible.

How that works ?
================

Compilertools dynamically sets link options and compiles options depending on the currently used compiler and targeted
architecture.

This avoids having to specify compiler-specific options in sources or setup files.

Multi-architecture optimized compilation for distribution
---------------------------------------------------------

**At build time:**

Compilertools helps to make optimized ".so"/".pyd" for each architecture and name files with tagged suffixes :

Example:

* *module.avx2.cp36-win_amd64.pyd* -> Optimized variant for AVX2 SIMD Extensions
* *module.avx.cp36-win_amd64.pyd* -> Optimized variant for AVX SIMD Extensions
* *module.cp36-win_amd64.pyd* -> Classic highly-compatible variant

These optimized files are packaged in the same wheel when distributing (Don't need to create a wheel by variant).

Requirement:

* Import ``compilertools.build`` before build module normally.
* Options can easily be changed directly in ``compilertools.build.CONFIG_BUILD`` dictionary.

**At runtime:**

Compilertools detects and chooses the best optimized ".so"/".pyd" to run based on CPU information.

If the best file not exists, search for the second best file, etc... If nothing found, use the highly-compatible one.

Requirement:

* Import ``compilertools`` one time before import-optimized modules (This adds a new import hook to Python).

Current-architecture optimized compilation
------------------------------------------

Compilertools finds the best compiler options for the current architecture and current compiler and build only one
optimized ".so"/".pyd" with classic name.

Requirement:

* Import ``compilertools.build`` before build module normally.

And also...
-----------

openMP, OpenACC, Intel Cilk Plus API auto-detection:
   Compilertools searches in source files for API ``pragma`` preprocessor calls and enables compiler and linker options
   if needed.

Extra generic compilers options:
   Compilertools can enable or disable generic extra compiler options like fast math.

Compatibility
=============

Supported Compilers
-------------------

Compilertools implements support for the following compilers:

* GCC
* LLVM Clang
* Microsoft Visual C++

Supported Processors
--------------------

Compilertools implements support for the following CPU:

* x86 (32 and 64 bits)

Build tools compatibility
-------------------------

Compilertools has been tested with the following build tools:

* Distutils
* Setuptools
* Numpy.distutils
* Cython


doc/changes.rst
Changelog
=========

1.1.5 (2023/05/05)
------------------

Others:

* Ensure Python 3.11 compatibility.
* Drop Python 3.6 and 3.7 support.

1.1.3 (2021/11/09)
------------------

Fixes:

 * Remove ``-mtune`` generic optimization for LLVM.

Others:

* Ensure Python 3.10 compatibility.

1.1.2 (2020/10/03)
------------------

Fixes:

 * Fix GCC/LLVM version detection in some cases.

Others:

* Use ``-O3`` instead of ``-O2`` on GCC/LLVM.
* Ensure Python 3.9 compatibility.
* Drop Python 3.5 support.

1.1.1 (2020/01/29)
------------------

Others:

* Ensure Python 3.8 compatibility.
* Drop Python 3.4 support.

1.1.0 (2019/01/11)
------------------

New compilers support:

* LLVM Clang

Others:

* Add more easy access function to get current machine CPU and compiler
  information.
* Use ``-O2`` instead of ``-O3`` on GCC.

1.0.0 (2018/05/27)
------------------

**First version with following features:**

Compilers:

* Microsoft Visual C++ compiler
* GNU Compiler Collection

Processors:

* X86 (32/64 bits)
* x86 CPUID

Others:

* Multi-architecture optimized compilation for distribution
* Current-architecture optimized compilation
* API autodetection: openMP, openACC & Intel Cilk Plus
* Extra compilers options: fast math
* Extended configuration
* Build tools compatibility: Distutils, Setuptools, Numpy.distutils & Cython
* PIP autodetection
* Full Unittest with pytest and Continuous integration on Appveyor and Travis-CI
* Sphinx Documentation with Readthedoc

Ideas & possibles futures features
----------------------------------

* Ability to force the use of an API (openMP, ...)
* Add it as an option in *setuptools* (or eventually merge it in *setuptools*).
* Since this module get many information on CPU, give possibility to users to
  access them easily as a dict.
* Add support for more compilers (Intel, LLVM, ...). The aim is to support all
  compilers available with *distutils* and *numpy.distutils*.
* Add support for more processors (ARM, ...).


compilertools/compilers/llvm.py
"""LLVM Clang."""


from compilertools.compilers import CompilerBase as _CompilerBase
from compilertools._utils import (
    dump_version as _dump_version,
    python_version as _python_version,
)

__all__ = ["Compiler"]


class Compiler(_CompilerBase):
    """LLVM Clang."""

    @_CompilerBase._memoized_property
    def option(self):
        """
        Compatibles Options.

        Returns
        -------
        dict
            Keys are options names, values are dict of arguments with keys in
            {'link', 'compile'}.
        """
        return {"fast_fpmath": {"compile": "-Ofast"}}

    @_CompilerBase._memoized_property
    def api(self):
        """
        Compatibles API.

        Returns
        -------
        dict
            Keys are API names, values are dict of arguments with keys in
            {'link', 'compile'}.
        """
        api = {}
        if self.version >= 3.7:
            api["openmp"] = {"compile": "-fopenmp", "link": "-fopenmp=libomp"}
        return api

    @_CompilerBase._memoized_property
    def version(self):
        """
        Compiler version used.

        Returns
        -------
        float
            Version.
        """
        if not self.current_compiler:
            return None
        return _dump_version("clang")

    @_CompilerBase._memoized_property
    def python_build_version(self):
        """
        Compiler version that was used to build Python.

        Returns
        -------
        float
            Version.
        """
        return _python_version("clang")

    def _compile_args_matrix(self, arch, cpu):
        """
        Return available Clang compiler options for the specified CPU architecture.

        Parameters
        ----------
        arch : str
            CPU Architecture.
        cpu : compilertools.processors.ProcessorBase subclass
            Processor instance

        Returns
        -------
        list of CompilerBase.Arg
            Arguments matrix.
        """
        # Generic optimisation
        args = [[self.Arg(args=["-flto", "-O3"])]]

        # Architecture-specific optimisations
        if arch == "x86_64":
            args += [
                # CPU Generic optimisations
                [self.Arg(args="-m64")],
                # CPU Instructions sets
                [
                    self.Arg(
                        args=["-mavx512cd", "-mavx512f"],
                        suffix="avx512",
                        import_if=(
                            "AVX512F" in cpu.features
                            and "AVX512CD" in cpu.features
                            and cpu.os_supports_xsave
                        ),
                        build_if=self.version >= 3.9,
                    ),
                    self.Arg(
                        args="-mavx2",
                        suffix="avx2",
                        import_if=("AVX2" in cpu.features and cpu.os_supports_xsave),
                    ),
                    self.Arg(
                        args="-mavx",
                        suffix="avx",
                        import_if=("AVX" in cpu.features and cpu.os_supports_xsave),
                    ),
                    self.Arg(),
                ],
            ]

        elif arch == "x86_32":
            args += [
                # CPU Generic optimisations
                [self.Arg(args="-m32")],
                # CPU Instructions sets
                [
                    self.Arg(
                        args=["-mfpmath=sse", "-mavx2"],
                        suffix="avx2",
                        import_if=("AVX2" in cpu.features and cpu.os_supports_xsave),
                    ),
                    self.Arg(
                        args=["-mfpmath=sse", "-mavx"],
                        suffix="avx",
                        import_if=("AVX" in cpu.features and cpu.os_supports_xsave),
                    ),
                    self.Arg(
                        args=["-mfpmath=sse", "-msse4"],
                        suffix="sse4",
                        import_if=(
                            "SSE4_1" in cpu.features and "SSE4_2" in cpu.features
                        ),
                    ),
                    self.Arg(
                        args=["-mfpmath=sse", "-msse4.2"],
                        suffix="sse4_2",
                        import_if="SSE4_2" in cpu.features,
                    ),
                    self.Arg(
                        args=["-mfpmath=sse", "-msse4.1"],
                        suffix="sse4_1",
                        import_if="SSE4_1" in cpu.features,
                    ),
                    self.Arg(
                        args=["-mfpmath=sse", "-msse4a"],
                        suffix="sse4a",
                        import_if=(
                            "SSE4A" in cpu.features and cpu.vendor == "AuthenticAMD"
                        ),
                    ),
                    self.Arg(
                        args=["-mfpmath=sse", "-mssse3"],
                        suffix="ssse3",
                        import_if="SSSE3" in cpu.features,
                    ),
                    self.Arg(
                        args=["-mfpmath=sse", "-msse2"],
                        suffix="sse2",
                        import_if="SSE2" in cpu.features,
                    ),
                    self.Arg(
                        args=["-mfpmath=sse", "-msse"],
                        suffix="sse",
                        import_if="SSE" in cpu.features,
                    ),
                    self.Arg(),
                ],
            ]

        return args

    def _compile_args_current_machine(self, arch, cpu):
        """
        Return auto-optimised Clang arguments for current machine.

        Parameters
        ----------
        arch : str
            CPU Architecture.
        cpu : compilertools.processors.ProcessorBase subclass
            Processor instance.

        Returns
        -------
        str
            Best compiler arguments for current machine.
        """
        args = ["-O3 -march=native -flto"]

        if arch == "x86_32":
            args.append("-m32")
            if "SSE" in cpu.features:
                args.append("-mfpmath=sse")

        elif arch == "x86_64":
            args.append("-m64")

        return " ".join(args)
